---
title: "Greer - H3K27me3 DiffBind Using hg38"
author: "Meeta Mistry"
date: "`r Sys.Date()`"
output:
   html_document:
      code_folding: hide
      highlights: pygments
      theme: default
      toc: true
      toc_float:
         collapsed: true
         smooth_scroll: true
---


## Overview

- Principal Investigator: Eric Greer
- Experiment: CUT&RUN analysis of microprotein over- and under-expression

## Project Description

**The difference between this report and the previous one, is the use of a more recent genome build (hg38).**

There are over 5000 predicted microproteins (< 10kDa) in human proteomes, which remain uncharacterized. From labeling experiments the lab has found a huge amount of signal from these small proteins

**Hypothesis:** these small proteins function as histone mimetics, recruiting chromatin modifiers away from the chromatin.

By evaluating homology of these proteins to histone tails, they identified Nop16 to have a high homology for H3K27. Nop16 over expression decreases H3K27me3 modification, and a knockout results in increase in H3K27me3.


```{r functions, echo=FALSE}
# Load functions
sanitize_datatable = function(df, ...) {
 # remove dashes which cause wrapping
 DT::datatable(df, ..., rownames=gsub("-", "_", rownames(df)),
                   colnames=gsub("-", "_", colnames(df)))
}

```

## Setup
### Bioconductor and R libraries used

```{r libraries, echo=TRUE}
loadlibs <- function(){

library(tidyverse)
library(reshape)
library(pheatmap)
library(RColorBrewer)
library(DT)
library(ggrepel)
library(viridis)
library(DiffBind)
library(magick)
library(cowplot)
# library(DESeq2)
}
suppressPackageStartupMessages(loadlibs())

```

## Reading in peaks and counting reads

DiffBind uses the alignment files (BAM) to compute count information for each of the peaks/regions in the consensus set. In this step, for each of the consensus regions DiffBind takes the number of aligned reads in the ChIP sample, to compute raw counts for each sample at every potential binding site. This is also done for the input samples separately. The peaks in the consensus peakset may be re-centered and trimmed based on calculating their summits (point of greatest read overlap) in order to provide more standardized peak intervals.

*NOTE: This part can be computationally extensive and so code described below was run on the O2 cluster and results saved to an R object.* 


```{r dbacount, eval=FALSE}
# For the O2 cluster
# module load gcc/6.2.0 R/4.1.1

# Point to the library location
.libPaths("/home/mm573/R/4.1.1")

# Load the library
library(DiffBind)

# Read in the samplesheet
samplesheet <- read.csv("DiffBind/samplesheet.csv")

# Create the db object
dbObj <- dba(sampleSheet=samplesheet)


# Figure out the ideal summits value
## You may want to look at your distribution of peak widths and choose a value accordingly, e.g. the minimum or first quartile value.
summary(dbObj$binding[,3]-dbObj$binding[,2])

# Calculate a binding matrix with scores based on read counts for every sample 
dbObj <- dba.count(dbObj, summits = 350, bParallel = FALSE)
```


```{r load-db, echo=FALSE}
# Load the Diffbind object
load("diffbind_analysis_hg38/dbObj_count.RData")
dbObj
```

## PCA and correlation heatmap {.tabset}
Here, we use DiffBind functions to plot the PCA and correlation heatmaps. For the 100K consensus regions the raw counts are input subtracted and a library-size normalization is set as the
default method within DiffBind. The RLE-based normalization (native DESeq2) and the TMM normalization (native edgeR), alters the data distribution to a greater extent. These methods developed for normalizing RNA-seq count matrices, produce results such that the binding changes are more evenly distributed between the
two conditions. However, we have no reason to expect changes to be balanced, and so we stick with the default.

**In both plots we observe a clear separation of samples by group.**

### PCA plot

```{r diffbind-pca}

# PCA using normalized bgsubtract counts
dba.plotPCA(dbObj,  attributes=DBA_FACTOR, label=DBA_ID,
            score = DBA_SCORE_NORMALIZED)

```

### Correlation Heatmap

```{r diffbind-corHeatmap}
# Correlation heatmap
dba.plotHeatmap(dbObj, ColAttributes = DBA_FACTOR,
                score = DBA_SCORE_NORMALIZED)
```

## Spike-in Normalization
An alternative normalization is to use spike-in data, where exogenous chromatin (in this case S. Cerevisiae) is "spiked in" to the ChIP. If the amount of spiked-in chromatin can be precisely controlled,
then we can use the relative amounts of reads that map to the alternative reference genome for each sample. DiffBind allows for spike-in reads to be included in the experiment as
an additional set of sequencing reads (BAM) files. We aligned all ChIP files to the sacCer3 genome and used those for input.

```{r spikein, eval=FALSE}
# Normalizing the data
dbObj <- dba.normalize(dbObj, spikein = TRUE)
```

```{r load-dbnorm, echo=FALSE}
# Load the Diffbind object
load("diffbind_analysis_hg38/dbObj_normalized.RData")
```

## Differential Enrichment Analysis
Every possible contrast (pair of comparisons) can be analyzed, but we provide below the two we are most interested in.

* **ng versus KO**
* **EV versus OE**

The `dba.analyze()` function by default will also run blacklist filtering and greylist filtering, both of which are briefly described below.

* **Blacklists:** are pre-defined lists of regions specific to a reference genome that are known to be problematic. The best known lists have been identified as part of the ENCODE project.
* **Greylists:** are specific to a ChIP-seq experiment, and are derived from the controls generated as part of the experiment. The idea is to identify anomalous regions where a disproportionate degree of signal is present. These regions can then be excluded from subsequent analysis.

_**Both of these filterings were previously applied in our bcbio pipeline, prior to peak calling.**_

By default DiffBind uses the DESeq2 package to identify differentially enriched regions. The spike in normalization will be applied rather than the RLE method internal to DESeq2.

```{r diffbind-de, progress=F}
# Establishing contrasts of interest
dbObj <- dba.contrast(dbObj, contrast = c("Factor", "EV", "OE"))
dbObj <- dba.contrast(dbObj, contrast = c("Factor", "ng", "KO"))

# Performing the differential enrichment analysis with DESeq2 default
dbObj <- dba.analyze(dbObj, method=DBA_DESEQ2,
                     bGreylist = FALSE, bBlacklist = FALSE)

# Summary of results
# Use padj (multiple test corrected) values to threshold
de_summary <- dba.show(dbObj, bContrasts=T, th=0.05)

sanitize_datatable(de_summary, style='bootstrap')
```

**These numbers are higher than what is reported for the previous analysis using GRCh37.**


## MA Plot {.tabset}

For each contrast, we can use the MA plot to evaluate regions that are being differentially enriched for direction of change and the overall enrichment.

_Each point represents a binding site, with points in magenta representing sites identified as differentially bound. There is a blue horizontal line through the origin (0 LFC), as well as a horizontal red curve representing a non-linear loess fit showing the underlying relationship between coverage levels and fold changes._

### Empty vector vs. Overexpression
A slightly larger number of magenta dots appear above the center line, indicating that most of the significant binding sites have a higher read count in the empty vector control samples. This suggests that the overexpression samples exhibit lower levels of the H3K27me3 mark. There are however, also sites that show the opposite effect.

```{r}
dba.plotMA(dbObj, method=DBA_DESEQ2, contrast = 1)
```

### ngRNA control versus Knockout
There are many more differentially enriched sites in this contrast and it's hard to say in which direction we observe the most change. Based on the loess curve, it appears that with coverage there is an increase in fold change (i.e. sites have higher read count in the ng control).


```{r}
dba.plotMA(dbObj, method=DBA_DESEQ2, contrast = 2)
```


## Annotate result files 
For four of the comparisons in which we identified significant differential enrichment of regions, we used ChIPseeker to find nearest gene annotations. The files listed below are .csv files which contain the statistics from the DE analysis and columns containing annotation information:

* [EV vs OE: total of 1485 regions](./results/combo_trb_spikeCtl_resAnnot.txt)
  * 1061 regions lose the H3K27 trimethylation mark in OE samples
  * 424 regions gain the H3K27 trimethylation mark in OE samples
* [ng vs KO: total of 3566 regions](./results/combo_dmso_spikeCtl_resAnnot.txt)
  * 2293 regions lose the H3K27 trimethylation mark in KO samples
  * 1273 regions gain the H3K27 trimethylation mark in KO samples

**These numbers are higher than what is reported for DiffBind, due to a one-to-many mapping from EntrezID to Ensembl.**

```{r anno-output, echo=FALSE, eval=FALSE, warning=FALSE, message=FALSE, progress=FALSE}
# Profile of ChIP peaks binding to TSS
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(AnnotationDbi)

# Get db
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# Create output files for all peaks (not just padj < 0.05)
for(cmp in c(1:2)){
  # Get results GRanges for contrast cmp
  res <- dba.report(dbObj, method=DBA_DESEQ2, contrast = cmp, th=1)
  
  # Annotate
  peakAnno <- annotatePeak(res, tssRegion=c(-3000, 3000), TxDb=txdb)
  
  # Add gene symbols, entrez Ids
library(EnsDb.Hsapiens.v86)
annotations_edb <- AnnotationDbi::select(EnsDb.Hsapiens.v86,
                keys = as.data.frame(peakAnno)$geneId,
                columns = c("SYMBOL", "GENEID"),
                keytype = "ENTREZID") 
annotations_edb$ENTREZID <- as.character(annotations_edb$ENTREZID)

# NOTE: Because hg38 is based on EntrezIds; there are entries in our results which have no gene symbol
  out <- peakAnno %>% 
    data.frame() %>% 
    dplyr::left_join(annotations_edb, by=c("geneId" = "ENTREZID"))
  filename <- paste0("results/", de_summary$Group[cmp], "_", de_summary$Group2[cmp], "_hg38_spikeCtl_resAnnot.txt")
  #write.csv(out, file=filename, quote=F, row.names=F)
  write_tsv(out, filename)
}
```

## Peak annotation overview {.tabset}
Here, we use different plots to visualize the genomic annotation.

```{r}
# Get sig diff peaks
oe_res <- dba.report(dbObj, contrast = 1, bUsePval = FALSE,  bCounts = TRUE, bCalled = T)
ko_res <- dba.report(dbObj, contrast = 2, bUsePval = FALSE,  bCounts = TRUE, bCalled = T)

# Separate by direction
lost_in_oe <- oe_res[(elementMetadata(oe_res)[, "Fold"] > 0)]

# Load libraries
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg38.knownGene)
library(AnnotationDbi)

# Get db
txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene

# Get annotations
peakAnno_oe <- annotatePeak(oe_res, tssRegion=c(-3000, 3000), TxDb=txdb)
peakAnno_ko <- annotatePeak(ko_res, tssRegion=c(-3000, 3000), TxDb=txdb)

```

### Feature Distribution
This plot identifies the percentage of total peaks which identify with the various genomic features. We can see that for both OE and KO, the majority of peaks found in distal intergenic or intronic regions.


```{r annobar}
# Plot figures
plotAnnoBar(peakAnno_oe)
plotAnnoBar(peakAnno_ko)
```

### Distribution of H3K27me3 marks relative to the TSS
The distance from the peak (binding site) to the TSS of the nearest gene is calculated by annotatePeak and reported in the output. In this plot we visualize the percentage of binding sites upstream and downstream from the TSS of the nearest genes. We observe a large percentage of binding sites that are changing to be quite far from the TSS (10-100KB or > 100KB).

```{r}
plotDistToTSS(peakAnno_oe, title="")
plotDistToTSS(peakAnno_ko, title="")
```

## Overlap with RNA-seq results {.tabset}
RNA-seq analysis was performed to evaluate transcriptional changes between the Nop16 KO and control, but also the Nop16 OE and control. Here, we identify overlaps between the RNA-seq results and the target genes from the differential enrichment analysis.

* Nop16 OE means that we should see a decrease in the H3K27me3 mark. We observe a majority of sites showing this trend. Because this is a loss of a repressive mark, we expect an increase in expression of those target genes.
* Nop16 KO means we should observe sites that increase in the H3K27me3 mark. For those genes that gain tri-methlyation in the KO, we would expect a decrease in expression due to the repressive nature of the mark.

_Note that ~20 of the target gene annotations in the KO and OE diffbind results are actually [LRG records](https://www.ensembl.info/2014/01/30/lrg-records-in-ensembl-stable-annotation-of-human-genes/), and will not have an overlap with RNA-seq data._

```{r, echo=FALSE, eval=FALSE}
## Output significant regions for overlap
oe <- read_tsv("diffbind_analysis_hg38/results/EV_OE_hg38_spikeCtl_resAnnot.txt")
ko <- read_tsv("diffbind_analysis_hg38/results/ng_KO_hg38_spikeCtl_resAnnot.txt")

oe %>% 
  dplyr::filter(FDR < 0.05 & Fold > 0) %>%  
  write_tsv("diffbind_analysis_hg38/results/H3k27me3_mark_lost_in_OE_sig.txt")

oe %>% 
  dplyr::filter(FDR < 0.05 & Fold < 0) %>%  
  write_tsv("diffbind_analysis_hg38/results/H3k27me3_mark_gained_in_OE_sig.txt")

ko %>% 
  dplyr::filter(FDR < 0.05 & Fold > 0) %>% 
  write_tsv("diffbind_analysis_hg38/results/H3k27me3_mark_lost_in_KO_sig.txt")

ko %>% 
  dplyr::filter(FDR < 0.05 & Fold < 0) %>% 
  write_tsv("diffbind_analysis_hg38/results/H3k27me3_mark_gained_in_KO_sig.txt")

```

```{r, message=FALSE, warning=FALSE}
library(VennDiagram)
# Read in appropriate RNA-seq files
genesUp_OE <- read.csv("../rnaseq_analysis_of_microprotein_hbc04238/results/sigGenes/OE_vs_control_sigUp_inOE.csv")
genesDown_KO <- read.csv("../rnaseq_analysis_of_microprotein_hbc04238/results/sigGenes/KO_vs_control_sigDown_inKO.csv")

# Read in the appropriate DiffBind results
k27me3lost_OE <- read.delim("diffbind_analysis_hg38/results/H3k27me3_mark_lost_in_OE_sig.txt") %>% drop_na(GENEID) %>% 
  pull(GENEID)
k27me3pos_KO <- read.delim("diffbind_analysis_hg38//results/H3k27me3_mark_gained_in_KO_sig.txt") %>%drop_na(GENEID) %>% 
  pull(GENEID)

```

### Nop16 Overexpression 

```{r venn-oe, eval=FALSE}
# Create lists for comparison
nop16oe <- list(
  k27targets = k27me3lost_OE,
  rnaseq = genesUp_OE$ensgene)

# Nop16oe
venn.diagram(x = nop16oe, 
             filename = "img/venn_hg38_nop16oe.png",
             output = TRUE,
             main = "Over-expression of Nop16",
             main.fontfamily = "sans",
             main.cex = 0.75,
             height = 600 , 
             width = 600 , 
             resolution = 300,
             compression = "lzw",
             lwd = 2,
             col= "gray",
             fill = c("salmon", "lightblue"),
             cex = 0.5,
             fontfamily = "sans",
             fontsize = 5,
             category.names = c("Genes decreased in \n H3K27me3 mark" , "Genes \n Up-regulated \n in OE"),
             cat.cex = 0.5,
             cat.fontface = "bold",
             cat.default.pos = "outer",
             cat.pos = 0,
             cat.dist = c(0.05,0.097),
             cat.fontfamily = "sans",
             cat.col = c("salmon", "lightblue"))
```


```{r venn-oe-img, out.width = '100%', message=F, warning=F}
img <- image_read("img/venn_hg38_nop16oe.png")
print(ggdraw() +
        draw_image(img))
```



### Nop16 Knockout

```{r venn-ko, eval=FALSE}
# Create lists for comparison
nop16ko <- list(
  k27targets = k27me3pos_KO,
  rnaseq = genesDown_KO$ensgene)

# Nop16ko
venn.diagram(x = nop16ko, 
             filename = "img/venn_hg38_nop16ko.png",
             output = TRUE,
             main = "Knockdown of Nop16",
             main.fontfamily = "sans",
             main.cex = 0.75,
             height = 600 , 
             width = 600 , 
             resolution = 300,
             compression = "lzw",
             lwd = 2,
             col= "gray",
             fill = c("salmon", "lightblue"),
             cex = 0.5,
             fontfamily = "sans",
             category.names = c("Genes \n increased in \n H3K27me3 mark" , "Genes Down-regulated \n in KO"),
             cat.cex = 0.5,
             cat.fontface = "bold",
             cat.default.pos = "outer",
             cat.pos = 180,
             cat.dist = c(0.05,0.05),
             cat.fontfamily = "sans",
             cat.col = c("salmon", "lightblue"))
```

```{r venn-ko-img}
img <- image_read("img/venn_hg38_nop16ko.png")
print(ggdraw() +
        draw_image(img))
```


## File Download
The following files from the overlap analysis are linked below:

**Nop16 Overexpression**

* Genes that lost the H3K27me3 mark _and_ up-regulated in RNA-seq data ([download](results/venn/OE_combined.tsv))
* Genes that lost the H3K27me3 mark but showed no corresponding increased expression in RNA-seq data ([download](results/venn/OE_k27me3_mark_lost_no_overlap_rnaseq.csv))
* Genes that were up-regulated in RNA-seq data but showed no differential enrichment of the H3k27me3 mark ([download](results/venn/OE_rnaseq_up_no_overlap_dbsig.csv))

**Nop16 Knockout**

* Genes that gained the H3K27me3 mark _and_ down-regulated in RNA-seq data ([download](results/venn/KO_combined.tsv))
* Genes that gained the H3K27me3 mark but showed no corresponding decreased expression in RNA-seq data ([download](results/venn/KO_k27me3_mark_gained_no_overlap_rnaseq.csv))
* Genes that were down-regulated in RNA-seq data but showed no differential enrichment of the H3k27me3 mark ([download](results/venn/KO_rnaseq_down_no_overlap_dbsig.csv))


```{r overlap-files, echo=FALSE, eval=FALSE}

## Overlaps first 
k27me3lost_OE %>% dplyr::inner_join(genesUp_OE,
                                 by=c("GENEID" = "ensgene")) %>% 
  write_tsv(file = "diffbind_analysis_hg38/results/venn/OE_combined.tsv")

k27me3pos_KO %>% dplyr::inner_join(genesDown_KO,
                                 by=c("GENEID" = "ensgene")) %>% 
  write_tsv(file = "diffbind_analysis_hg38/results/venn/KO_combined.tsv")

## Then those that didn't overlap
## 1. Load full files
rnaseqAll_oe <- read.csv("../rnaseq_analysis_of_microprotein_hbc04238/results/OE_vs_control_res_allgenes.csv")
rnaseqAll_ko <- read.csv("../rnaseq_analysis_of_microprotein_hbc04238/results/KO_vs_control_res_allgenes.csv")

k27db_oe <- read.delim("diffbind_analysis_hg38/results/EV_OE_hg38_spikeCtl_resAnnot.txt")
k27db_ko <- read.delim("diffbind_analysis_hg38/results/ng_KO_hg38_spikeCtl_resAnnot.txt")

# Read in DiffBind files
k27me3lost_OE <- read.delim("diffbind_analysis_hg38/results/H3k27me3_mark_lost_in_OE_sig.txt")
k27me3pos_KO <- read.delim("diffbind_analysis_hg38/results/H3k27me3_mark_gained_in_KO_sig.txt")

## 2. Now get the appropriate overlaps (OE)
k27me3lost_only <- k27me3lost_OE[which(k27me3lost_OE$GENEID %in% genesUp_OE$ensgene == FALSE),] %>% 
  left_join(rnaseqAll_oe, by=c("GENEID" = "ensgene"))

genesUp_OE_only <- genesUp_OE[which(genesUp_OE$ensgene %in% k27me3lost_OE$GENEID == FALSE),] %>% 
  left_join(k27db_oe, by=c("ensgene" = "GENEID"))

## 3. Now get the appropriate overlaps (KO)
k27me3gained_only <- k27me3pos_KO[which(k27me3pos_KO$GENEID %in% genesDown_KO$ensgene == FALSE),] %>% 
  left_join(rnaseqAll_ko, by=c("GENEID" = "ensgene"))

genesDown_KO_only <- genesDown_KO[which(genesDown_KO$ensgene %in% k27me3pos_KO$GENEID == FALSE),] %>% 
  left_join(k27db_ko, by=c("ensgene" = "GENEID"))

## 4. Write to file

write.csv(k27me3lost_only, 
          file = "diffbind_analysis_hg38/results/venn/OE_k27me3_mark_lost_no_overlap_rnaseq.csv", 
          quote = F, row.names = F)

write.csv(k27me3gained_only, 
          file = "diffbind_analysis_hg38/results/venn/KO_k27me3_mark_gained_no_overlap_rnaseq.csv", 
          quote = F, row.names = F)

write.csv(genesUp_OE_only, 
          file = "diffbind_analysis_hg38/results/venn/OE_rnaseq_up_no_overlap_dbsig.csv", 
          quote = F, row.names = F)

write.csv(genesDown_KO_only, 
          file = "diffbind_analysis_hg38/results/venn/KO_rnaseq_down_no_overlap_dbsig.csv", 
          quote = F, row.names = F)

```



## SessionInfo
A detailed summary of all the packages and package versions used to create this report.

```{r sessionInfo}

sessionInfo()
```
